vernac/comDefinition.ml

```diff
@@ -48,57 +48,59 @@ let warn_implicits_in_term =
-        let evd, (c, imps2) = interp_casted_constr_evars_impls ~impls env_bl evd c ty in
-        let body = EConstr.(it_mkLambda_or_LetIn c ctx) in
-        let typ = EConstr.(it_mkProd_or_LetIn ty ctx) in
-        (* 2 *)
-        let beq b1 b2 = if b1 then b2 else not b2 in
-        let impl_eq (x,y,z) (x',y',z') = beq x x' && beq y y' && beq z z' in
-        (* Check that all implicit arguments inferable from the term
-           are inferable from the type *)
-        let chk (key,va) =
-          impl_eq (List.assoc_f Pervasives.(=) key impsty) va (* FIXME *)
-        in
-        if not (try List.for_all chk imps2 with Not_found -> false)
-        then warn_implicits_in_term ();
-        (* 3 *)
-        let evd, _ = Evarutil.nf_evars_and_universes evd in
-        let bodyvars = EConstr.universes_of_constr env evd body in
-        let tyvars = EConstr.universes_of_constr env evd ty in
-        let vars = Univ.LSet.union bodyvars tyvars in
-        let evd = Evd.restrict_universe_context evd vars in
-        let uctx = Evd.check_univ_decl ~poly evd decl in
-        (* 4 *)
-        let body,typ = EConstr.(to_constr evd body, to_constr evd typ) in
-        evd, imps1@(Impargs.lift_implicits nb_args impsty),
-          definition_entry ~types:typ ~univs:uctx body
+  (* Build the type *)
+  let evd, tyopt =
+    Option.cata (fun ctyp ->
+        let evd, ty = interp_type_evars_impls ~impls env_bl evd ctyp in
+        evd, Some ty)
+      (evd, None) ctypopt
+  in
+  (* Build the body, and merge implicits from parameters and from type/body *)
+  let evd, c, imps, tyopt =
+    match tyopt with
+    | None ->
+      let evd, (c, impsbody) = interp_constr_evars_impls ~impls env_bl evd c in
+      evd, c, imps1@Impargs.lift_implicits (Context.Rel.nhyps ctx) impsbody, None
+    | Some (ty, impsty) ->
+      let evd, (c, impsbody) = interp_casted_constr_evars_impls ~impls env_bl evd c ty in
+      check_imps ~impsty ~impsbody;
+      evd, c, imps1@Impargs.lift_implicits (Context.Rel.nhyps ctx) impsty, Some ty
+  in
+  (* universe minimization *)
+  let evd = Evd.nf_constraints evd in
+  (* Unless program mode the body and type are ready, substitute evars *)
+  let body = to_constr evd (it_mkLambda_or_LetIn c ctx) in
+  let tyopt = Option.map (fun ty -> to_constr evd (it_mkProd_or_LetIn ty ctx)) tyopt in
+  (* Keep only useful universes (NB: if program mode evars may remain) *)
+  let bodyvars = universes_of_constr env evd (of_constr body) in
+  let tyvars =
+    Option.cata (fun ty -> universes_of_constr env evd (of_constr ty))
+      Univ.LSet.empty tyopt
   in
+  let vars = Univ.LSet.union bodyvars tyvars in
+  let evd = Evd.restrict_universe_context evd vars in
+  (* Check we conform to declared universes *)
+  let uctx = Evd.check_univ_decl ~poly evd decl in
+  (* We're done! *)
+  let ce = definition_entry ?types:tyopt ~univs:uctx body in
   (red_constant_entry (Context.Rel.length ctx) ce evd red_option, evd, decl, imps)
```

Comment:
Something that puzzles me is that we return a ground definition here, but yet the evar_map and the universe context is live. So the term is maybe not ground?

IMHO, if that's the case it could be better to delay the `to_constr` call, but I dunno.

Reply:
This is the program_mode case. I don't really understand program so I don't know what's feasible there.

Reply:
Well, `evd` is used also once in the non-program case `Evd.universe_binders evd` so maybe that should be also factored out so indeed we could wrap all the program-specific info into a Program structure.